torch/csrc/jit/tensorexpr/loopnest.cpp

```diff
@@ -667,10 +683,11 @@ class FunctionInliner : public IRMutator {
     if (v == producer_ && !outputs_.count(buf_)) {
       in_producer_ = true;
       producer_ = to<Store>(IRMutator::mutate(v));
-      TORCH_INTERNAL_ASSERT(
-          producer_,
-          buildErrorMessage(
-              "Producer statement for output buf should remain non-null in the fuser"));
+      if (!producer_) {
+        // Producer statement for output buf should remain non-null in the fuser
+        success_ = false;
+        return v;
+      }
```

Comment:
Hm, what will happen if inliner failed somewhere on loads? Are we still going to delete the producing statement?

Since we're now performing mutations in place I wonder if this all can together lead to a partially updated, and hence incorrect, IR.

Reply:
Actually, there are few complications here, which I didn't fully think through. One is that the producer statement will be mutated before the loads, so at this point there is no way to know if the inlining would succeed. Given this, the logic of returning a flag would never work.

Also, good point regarding partially successful inlining. The only way around these is to clone the IR and do what Vectorizer is doing. But then we will lose the in-place mutation property for inlining. I'm not sure which is better.

Reply:
Btw, I think that my recent similar change to vectorizer might have a similar problem too, I need to double check. It would be good to split analysis and transformation parts in these passes, like good ol' guidelines from compiler development tell us :)

Reply:
Yeah, it is better to check the vectorizer as well. 

We should definitely split the analysis and transformation phases. One easy way to do that in our current setup is to do what loop fusion is doing now; clone the IR, try inlining the cloned IR, if it fails return the old IR, if it succeeds, then redo the inlining on the old IR.

Reply:
Right, that sounds like a sturdy solution until we implement a solid single-pass one.

Reply:
Updated this PR as per this discussion. PTAL.